proposals/2024-06-05_Track_Similarity_Column.md

```diff
@@ -0,0 +1,168 @@
+
+def pitch_diff(key1, bpm1, key2, bpm2):
+  # calculate the pitch delta for each track when stretched to a tempo of 100BPM
+  delta1 = pitch_delta_on_tempo_stretching(bpm1, 100)
+  delta2 = pitch_delta_on_tempo_stretching(bpm2, 100)
+
+  # get the resulting key for each track at 100BPM
+  res_key1 = normalize_pitch_value(key1 + delta1)
+  res_key2 = normalize_pitch_value(key2 + delta2)
+
+  # now return the pitch difference when both tracks are played at the same tempo
+  return normalize_pitch_value(res_key1 - res_key2)
+
+pd = pitch_diff(key1, bpm1, key2, bpm2)
+print("Pitch Difference:", pd) # Output: 0.19895410624553378
+
+```
+
+For the examples above, we find that both these tracks are almost in the same key (with a 20 cent difference)
+when they’re played at the same tempo.We would not have known this by looking at the keys or the BPMs independently.
+This method of calculating harmonic compatibility gives users a deeper insight on which tracks will work well.
+
+Harmonically compatible tracks do not need to end up in the same key. We can account for this using the Circle of Fifths.
+Here’s the continued Python pseudocode:
+
+```python
+circle_of_fifths = [
+    0,  # C
+    7,  # G
+    2,  # D
+    9,  # A
+    4,  # E
+    11,  # B
+    6,  # Gb
+    1,  # Db
+    8,  # Ab
+    3,  # Eb
+    10,  # Bb
+    5,  # F
+]
+
+def circle_of_fifths_distance(note):
+    # distance of given note from note 0
+    # it's a circle, so the shortest distance may be from the end of the array
+    return min(circle_of_fifths.index(note), 12 - circle_of_fifths.index(note))
+
+def compatibility(pitch_difference):
+    rounded_pitch_diff = round(pitch_difference)
+    delta = abs(rounded_pitch_diff - pitch_difference)
+    # get a rank from 0 to 6, lower means more compatible
+    cof_rank = circle_of_fifths_distance(rounded_pitch_diff)
+    return cof_rank, delta
+
+comp = compatibility(pd)
+print("Rank:", comp[0], "Delta:", comp[1]) # Rank: 0 Delta: 0.19895410624553378
+```
+
+This gives us a rank, where 0 means the tracks will be in the same key,
+and 6 means the tracks will sound most dissonant together.
+
+We also get a delta/detune value to find out how off tune the tracks will be in cents.
+
+TODO: The algorithm needs to be able to take in tuning information, for non 440Hz tuning standards.
+
+### Displaying The Information
+
+A similarity column will show the rank.  The column will allow for sorting and ranking search results.
+The values may be encoded as colors taken from a fire, in the range of black to bright yellow.
+
+Detune value may be displayed in another column. It ranges from 0 to 0.5. Lower values mean more harmonious transitions.
+
+**Open Question**: It may be possible to combine the information of the detune column into the similarity column.
+Either by creating a formula that combines the two values,
+or by encoding it in some other way (like opacity, or another icon).
```

Comment:
Since the Detune value destroys the assumptions in the similarity column, it should be directly considered in the value of the similarity column. We may consider to add controls to see the detune in the decks to all skins not have a Shade Skin only feature. Putting this into the similarity column as an extra value is probably confusing.